File: chucknorris.py
#!/usr/local/bin/python3

import requests, json, datetime, time
from pprint import pprint


# ================================ Chuck Norris IO SMS Engine ================================

# Description:
# Welcome to Python Chuck Norris IO SMS Engine. The purpose of this PY is to serve as 
# SMS Engine to broadcast the daily Chuck Norris Facts to a certain destination.
# This is a piece of the SMS Broadcast system for Chuck Norries facts. Remember: Chuck Norries
# does not need an API, he IS the API. 

# Author: 
# Filipe Leitão (contact@fleitao.org)

# Application:
appName = "chucknorris"

# External Service/API to be used: Chuck Norris Quote
chuck_api = "https://api.chucknorris.io/jokes/random"

# Timestamp for logs and debug
timest = datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d %H:%M:%S')

# ============================================================================================



###############################################################################
#                         CPaaS Provider Data / Restcomm
###############################################################################

rc_protocol       = "https"
rc_server         = "<<your_restcomm_domain>>";
rc_restCommBase   = "restcomm";
rc_accountBase    = "2012-04-24/Accounts";
rc_accountSid     = "<<your_restcomm_accountSID>>";
rc_accountToken   = "<<your_restcomm_accountToken>>";
rc_application    = "SMS/Messages.json";
      
rc_url = rc_protocol
rc_url = rc_url + '://'
rc_url = rc_url + rc_server
rc_url = rc_url + '/'
rc_url = rc_url + rc_restCommBase
rc_url = rc_url + '/'
rc_url = rc_url + rc_accountBase
rc_url = rc_url + '/'
rc_url = rc_url + rc_accountSid
rc_url = rc_url + '/'
rc_url = rc_url + rc_application

# rc_url is a clean Restcomm url ready to consume payload

# For Debug - uncomment as needed
print ("["+timest+"] DEBUG - Restcomm URL: " + rc_url)


#######################  Get Chuck Norris Quote as JSON #######################  
#
# Typical Response Format:
#
# {'category': None,
#  'icon_url': 'https://assets.chucknorris.host/img/avatar/chuck-norris.png',
#  'id': 'Ziyk6Vi5TqCscqxWsuR8Lg',
#  'url': 'https://api.chucknorris.io/jokes/Ziyk6Vi5TqCscqxWsuR8Lg',
#  'value': 'Chuck Norris once started a fire with only what was around him. '
#           'He was on an iceberg.'}
#
###############################################################################

# Get Response from API Request
resp = requests.get(chuck_api)

# Decode Response as JSON
resp_json = resp.json()

# For Debug - uncomment as needed
print ("["+timest+"] DEBUG - JSON Response Received:")
pprint(resp_json)

# Get Fact from Response
cn_fact = resp_json['value']

# For Debug - uncomment as needed
print ("["+timest+"] DEBUG - Quote Received: " + cn_fact)

                   
##################### Chuck Norris Specific Code Ends Here #################### 



###############################################################################
#                  SMS Broadcast Engine - Restcomm Based
###############################################################################

# Build JSON file path to be loaded
json_path = "/home/ubuntu/workspace/" 
json_path = json_path + appName
json_path = json_path + "/" 
json_path = json_path + appName 
json_path = json_path + ".json"

# Load JSON file with the users list
with open(json_path) as f:
    user_list_json = json.load(f)

# Initialize numbers list (just for debug)
numbers_list = ""

# Goes over the list and sends message to each number
for number in user_list_json:
    cn_payload = {'To': number, 'From': 'ChuckFacts', 'Body': cn_fact}
    from requests.auth import HTTPBasicAuth
    resp_sms = requests.post(rc_url, auth=(rc_accountSid,rc_accountToken), data=cn_payload)
    numbers_list = numbers_list + " " + number

# For Debug - uncomment as needed
print ("["+timest+"] DEBUG - Message Sent To: "+numbers_list) 

File: chucknorris.js

// ================================ Chuck Norris IO Middleware ================================

// Description:
// Welcome to Node.JS Chuck Norris IO Middleware. The purpose of this JS is to serve as 
// middleware between an SMS Request Factory, like restcomm.com and api.chucknorriasd.io.
// This is a piece of the SMS Broadcast system for Chuck Norries facts. Remember: Chuck Norries
// does not need an API, he IS the API. 

// Author: 
// Filipe Leitão (contact@fleitao.org)

// Application:
const appName = "chucknorris";

// ============================================================================================


var express = require('express');
var fs = require("fs");
var request = require('request');


// =============== Restcomm Account Details =============== 

  var rc_server         = "<<your_restcomm_domain>>";
  var rc_restCommBase   = "restcomm";
  var rc_accountBase    = "2012-04-24/Accounts";
  var rc_accountSid     = "<<your_restcomm_accountSID>>";
  var rc_accountToken   = "<<your_restcomm_accountToken>>";
  var rc_application    = "SMS/Messages";
  
  var rc_path =  rc_restCommBase + '/' 
                + rc_accountBase + '/'  
                + rc_accountSid + '/' 
                + rc_application;

const serviceName       = "ChuckFacts"

//console.log("[LOG] DEBUG - Restcomm API Path Loaded: %s",rc_path);

// =============== RESTful API Creation =============== 

var rest = express();



// =============== RESTful POST API ADD SUBSCRIBER  =============== 


// GET - Add User: https://<<your_api_domain>>/chucknorris/list?number=123456789&newUser=true

// RETURNS:
//  - 0: Not Authorised
//  - 1: Message Broadcasted
//  - 2: Subscription Successful
//  - 3: User Exists

rest.get('/'+appName+'/list', function (req, res) {
    

    // get timestamp
    var now = new Date();
    var timestamp = 'utc|' + now.getUTCFullYear() 
                           + '/' + (now.getUTCMonth()+1)
                           + '/' + now.getUTCDate()
                           + '|' + now.getHours()
                           + ':' + now.getMinutes();


    // console.log("[%s] DEBUG - Request Content: %s",timestamp,req);


   // First read existing users.
   fs.readFile( "<<path>>/"+appName+"/"+appName+".json", 'utf8', function (err, data) {
       if (err) {
            return console.error(err.message);
         }
    
        var obj = {
           subscribers: []
        };
        

        obj = JSON.parse(data); //conver it to an object

        //define SMS origin number (to add)
        var originNumber = req.query.number;
        var destNumber = req.query.dest;
        var newUser = req.query.newUser;
        var listAll = req.query.listAll;


        // if newUser = true, then it's a new subscription attempt
        if (newUser=="true"){
    
            //  double-check if number exists
            if(obj.hasOwnProperty(originNumber)){
                console.log("[%s] SERVER - Double subscription attempt: %s - %s",timestamp, originNumber);
                res.end('3');
            } else{
            
                //add number
                obj[originNumber] = {number:originNumber,created:timestamp,dest:destNumber};
        
                var json = JSON.stringify(obj); //convert it back to json
        
               fs.writeFile("<<path>>/"+appName+"/"+appName+".json", json, 'utf8', function (err, data) {
                    if (err) {
                        return console.error(err.message);
                     }
                    
                    console.log("[%s] SERVER - new subscriber added: %s ",timestamp,originNumber);
                    res.end('2');
                });
            }

        // if listAll = true, then it's time to list them all
        } else if(listAll=="true"){
            
            fs.readFile( "<<path>>/"+appName+"/"+appName+".json", 'utf8', function (err, data) {
                if (err) {
                    return console.error(err.message);
                }
            
            console.log("[%s] SERVER - listall successful", timestamp);
            
            res.end( data );

            });
            
        } 

    });
})



// =============== RESTful API DELETE SUBSCRIBER =============== 

// DELETE - https://<<your_api_domain>>/chucknorris/delete?number=123456789

// RETURNS:
//  - 1: User Deleted


rest.delete('/'+appName+'/delete', function (req, res) {
    
        // get timestamp
    var now = new Date();
    var timestamp = 'utc|' + now.getUTCFullYear() 
                       + '/' + (now.getUTCMonth()+1)
                       + '/' + now.getUTCDate()
                       + '|' + now.getHours()
                       + ':' + now.getMinutes();

    
    fs.readFile( "<<path>>/"+appName+"/"+appName+".json", 'utf8', function (err, data) {
        if (err) {
            return console.error(err.message);
         }
        
        var obj = JSON.parse(data);

        //define sms origin number
        var originNumber = req.query.number;

        // delete number
        delete obj[originNumber];
        
        var json = JSON.stringify(obj); //convert it back to json

       fs.writeFile("<<path>>/"+appName+"/"+appName+".json", json, 'utf8', function (err, data) {
            if (err) {
                return console.error(err.message);
             }
            console.log("[%s] SERVER - subscriber deleted: %s",timestamp,originNumber);
            res.end('1');
        });    
    })
})



// =============== RESTful Server Start =============== 

var server = rest.listen(8081, function () {

  var host = server.address().address
  var port = server.address().port

  console.log("[LOG] SERVER - "+appName+" app listening at http://%s:%s", host, port)

})
